providers/common/ciphers/tdes.c

```diff
@@ -0,0 +1,430 @@
+static OSSL_OP_cipher_ctx_get_params_fn tdes_ctx_get_params;
+static OSSL_OP_cipher_ctx_set_params_fn tdes_ctx_set_params;
+
+static int cipher_get_params(OSSL_PARAM params[], int mode,
+                             unsigned long int flags, size_t kbits,
+                             size_t blkbits, size_t ivbits)
+{
+    OSSL_PARAM *p;
+
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_MODE);
+    if (p != NULL && !OSSL_PARAM_set_int(p, mode)) {
+        PROVerr(0, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_FLAGS);
+    if (p != NULL && !OSSL_PARAM_set_ulong(p, (flags))) {
+        PROVerr(0, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_KEYLEN);
+    if (p != NULL && !OSSL_PARAM_set_int(p, (kbits) / 8)) {
+        PROVerr(0, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_BLOCK_SIZE);
+    if (p != NULL && !OSSL_PARAM_set_int(p, (blkbits) / 8)) {
+        PROVerr(0, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_IVLEN);
+    if (p != NULL && !OSSL_PARAM_set_int(p, (ivbits) / 8)) {
+        PROVerr(0, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    return 1;
+}
+
+static void *tdes_newctx(void *provctx, int mode, size_t kbits,
+                         const PROV_TDES_CIPHER *ciph)
+{
+    PROV_TDES_CTX *ctx = OPENSSL_zalloc(sizeof(*ctx));
+
+    ctx->mode = mode;
+    ctx->keylen = kbits / 8;
+    ctx->pad = 1;
+    ctx->ciph = ciph;
+    ctx->libctx = PROV_LIBRARY_CONTEXT_OF(provctx); /* used for rand */
+    return ctx;
+}
+
+static void tdes_freectx(void *vctx)
+{
+    PROV_TDES_CTX *ctx = (PROV_TDES_CTX *)vctx;
+
+    OPENSSL_clear_free(ctx,  sizeof(*ctx));
+}
+
+static int tdes_init(void *vctx, const unsigned char *key, size_t keylen,
+                     const unsigned char *iv, size_t ivlen, int enc)
+{
+    PROV_TDES_CTX *ctx = (PROV_TDES_CTX *)vctx;
+
+    ctx->enc = enc;
+
+    if (iv != NULL) {
+        if (ivlen != TDES_IVLEN) {
+            PROVerr(0, PROV_R_INVALID_IVLEN);
+            return 0;
+        }
+        memcpy(ctx->iv, iv, TDES_IVLEN);
+    }
+
+    if (key != NULL) {
+        if (keylen != ctx->keylen) {
+            PROVerr(0, PROV_R_INVALID_KEYLEN);
+            return 0;
+        }
+        return ctx->ciph->setkey(ctx, key, ctx->keylen);
+    }
+    return 1;
+}
+
+static int tdes_einit(void *vctx, const unsigned char *key, size_t keylen,
+                     const unsigned char *iv, size_t ivlen)
+{
+    return tdes_init(vctx, key, keylen, iv, ivlen, 1);
+}
+
+static int tdes_dinit(void *vctx, const unsigned char *key, size_t keylen,
+                      const unsigned char *iv, size_t ivlen)
+{
+    return tdes_init(vctx, key, keylen, iv, ivlen, 0);
+}
+
+#define BLOCK_SIZE DES_BLOCK_SIZE
```

Comment:
Why this?

Reply:
Hmm - well I am pondering whether to make this function a macro since it is exactly the same as the aes code (apart from the block_size and the ctx).

Reply:
blocksize is gone now